### PR TITLE
Added a SocialAuth.Middleware.afterLogin event.

### PR DESCRIPTION
**Pull request by joshrhykerd**

### CODE REVIEW

File: src/Middleware/SocialAuthMiddleware.php [hunk 1]
```diff
@@ -30,6 +30,9 @@ class SocialAuthMiddleware
     use InstanceConfigTrait;
     use ModelAwareTrait;
 
+    /** @const string The name of the event that is fired immediately afterLogin. */
```
Comment (ADmad) [inline src/Middleware/SocialAuthMiddleware.php]: Please use multi line docblock.

File: src/Middleware/SocialAuthMiddleware.php [hunk 2]
```diff
@@ -30,6 +30,9 @@ class SocialAuthMiddleware
...
+    const EVENT_AFTER_LOGIN = 'SocialAuth.Middleware.afterLogin';
```
Comment (ADmad) [inline src/Middleware/SocialAuthMiddleware.php]: Just `SocialAuth.afterLogin` is good enough.

File: src/Middleware/SocialAuthMiddleware.php [hunk 3]
```diff
@@ -182,6 +185,8 @@ protected function _handleCallbackAction(ServerRequest $request, Response $respo
 
         $request->session()->write($config['sessionKey'], $user);
 
+        $event = $this->_userModel->dispatchEvent(self::EVENT_AFTER_LOGIN, ['user' => $user]);
```
Comment (ADmad) [inline src/Middleware/SocialAuthMiddleware.php]: `$event` variable is unused so no need to assign it.

**Comment by joshrhykerd** [comment]
Currently I'm only using it to trigger an update of the last_login time in the database. I have thought of another way to do this, but I thought events would make more sense.

**Comment by ADmad** [comment]
Thats a reasonable use case. 

Looking good. I am just concerned if it's appropriate to dispatch the event on model instance. Model layer should not be concerned by about login.

**Comment by joshrhykerd** [comment]
I couldn't find a way to get ahold of the middleware object in order to attach event handlers directly to it.

**Comment by joshrhykerd** [comment]
https://github.com/ADmad/cakephp-social-auth/pull/26/commits/79f3b9113a16e69517109eff54133abe5a65176b is just a suggestion. In the docs `userEntity` is defined as **Boolean indicating whether user identity should be returned as entity.** Well if this boolean is switched to true then not only is an entity returned, but the session data also gets written as the entity instead of an array. I think as its defined in the docs it sounds like only the return value should be being modified. I actually have the user case right now where I want the user session data to be an array like normal, but I'd like to receive an entity back.

File: src/Middleware/SocialAuthMiddleware.php [hunk 4]
```diff
@@ -176,11 +183,13 @@ protected function _handleCallbackAction(ServerRequest $request, Response $respo
 
         $user->unsetProperty($config['fields']['password']);
 
+        $request->session()->write($config['sessionKey'], $user);
+
         if (!$config['userEntity']) {
             $user = $user->toArray();
         }
 
-        $request->session()->write($config['sessionKey'], $user);
```
Comment (ADmad) [inline src/Middleware/SocialAuthMiddleware.php]: Why did you move up this statement? This would cause the session record to always be entity make the `userEntity` config redundant.

**Comment by ADmad** [comment]
You should instead move the event triggering above the 'userEntity'  config check so that event handler always gets an entity.

File: src/Middleware/SocialAuthMiddleware.php (hunk 4 above)
Comment (joshrhykerd) [reply src/Middleware/SocialAuthMiddleware.php]: I explained my reasoning above, but my reasoning does not exactly match the code change that I made lol. I will revert this, but look at what I stated above and see if it makes sense for some type of small change here.

**Comment by joshrhykerd** [comment]
I think this is what I meant to do, but I accidentally moved the wrong thing in haste. Is this is change that makes sense and would you like me to do it?

File: src/Middleware/SocialAuthMiddleware.php (hunk 4 above)
Comment (joshrhykerd) [reply src/Middleware/SocialAuthMiddleware.php]: I guess what we need to determine is exactly what `userEntity` is trying to control. The value stored in the session or value sent in the event triggering or both.

**Comment by ADmad** [comment]
Please checkout how [GlideMiddleware](https://github.com/ADmad/cakephp-glide/blob/master/src/Middleware/GlideMiddleware.php) implement event dispatching and implement similar.

You can then use the global event manager to attach a listener for the afterLogin event anywhere you want in your app.

**Comment by joshrhykerd** [comment]
We've discussed in short whether or not it's appropriate to dispatch the event on model instance. Is this a change you would like me to implement? to dispatch the events on the SocialAuth middleware itself instead?

**Comment by joshrhykerd** [comment]
Are you referring mainly to fact that the event is dispatched on the middleware itself instead of the model or is there something else I am also missing?

File: src/Middleware/SocialAuthMiddleware.php (hunk 4 above)
Comment (joshrhykerd) [inline src/Middleware/SocialAuthMiddleware.php]: This commit was reverted in d9598f31933521fa11b2df3af108994dce6f68b9 as requested.

**Comment by BassemN** [comment]
@ADmad I have the same issue, the event `Auth.afterIdentify` is no longer fired after login like the old plugin

**Comment by ADmad** [comment]
This is now implemented on the `dev` branch. `SocialAuth.afterIdentify` event is now dispatched upon successful login.

**Comment by BassemN** [comment]
Thanka allot @ADmad When we can expect a tagged release?

**Comment by ADmad** [comment]
When I get time to test out the `dev` branch or some one give feedback for it :)

**Comment by BassemN** [comment]
I'll try to test it at my end and feed you back. Will you add new features to the `dev` branch?

**Comment by ADmad** [comment]
Don't have any plans to add any more new features to `dev` for now.

**Comment by BassemN** [comment]
I tested the following on my App and evrything is working well
- Make social profile model configurable.
- Dispatch SocialAuth.afterIdentify event.
